### ISSUE TITLE
Package matcher with identity attribute not working with apache2 cookbook?

### ISSUE THREAD

**Issue by tknerr** [sethvargo/chefspec#582]
I see that [`expect(chef_run).to install_package('identity_attribute')` should work](https://github.com/sethvargo/chefspec/blob/8e496bd28b59cfc1d3f169f421265d03341b4fe7/examples/package/spec/install_spec.rb#L21), and i just ran it to verify that it passes. And it does.

However, it fails with an example where I have `include_recipe` the apache2 cookbook, which uses [this package resource](https://github.com/svanzoest-cookbooks/apache2/blob/v3.0.1/recipes/default.rb#L20-22) in the default recipe:
```ruby
package 'apache2' do
  package_name node['apache']['package']
end
```

In [my chefspec test](https://github.com/tknerr/sample-toplevel-cookbook/blob/update-gems/test/chefspec/default_spec.rb) I'm using this:
```ruby
  context "on ubuntu systems" do
    subject { chef_run(platform: 'ubuntu', version: '12.04') }
    it { should install_package 'apache2' }
    it { should start_service 'apache2' }
  end

  context "on centos systems" do
    subject { chef_run(platform: 'centos', version: '6.3') }
    it { should install_package 'httpd' }
    it { should start_service 'httpd' }
  end
```

On latest state in ["master"](https://github.com/tknerr/sample-toplevel-cookbook/tree/d83a8a3fedd923ca8abc9e3f68aaeac803255402) everything passes:
https://circleci.com/gh/tknerr/sample-toplevel-cookbook/143

On the ["update-gems" branch](https://github.com/tknerr/sample-toplevel-cookbook/tree/update-gems), which involves a Chefspec 4.1.1 -> 4.2.0 and Chef 11.16.4 -> 12.3.0.dev update, the test for `it { should install_package 'httpd' }` suddenly fails for the `centos` platform:
https://circleci.com/gh/tknerr/sample-toplevel-cookbook/144:
```
rspec -fd --color -I test/chefspec test/chefspec

sample-app::default
  on ubuntu systems
    should install package "apache2"
    should start service "apache2"
  on centos systems
    should install package "httpd" (FAILED - 1)
    should start service "httpd"
  default 'words_of_wisdom'
    should render file "/var/www/sample.html" with content "I can haz cats!"
  when 'words_of_wisdom' are set
    should render file "/var/www/sample.html" with content "no cats today :-("
  with data bag items
    should render file "/var/www/sample.html" with content "<li>ananas: yellow & sweet</li>"
  when data_bag_path is invalid
    should write log "can not load data_bag: Chef::Exceptions::InvalidDataBagPath"
    should render file "/var/www/sample.html" with content /<ul>\s*<\/ul>/

Failures:

  1) sample-app::default on centos systems should install package "httpd"
     Failure/Error: it { should install_package 'httpd' }
       expected "package[httpd]" with action :install to be in Chef run. Other package resources:
       
         
        
     # ./test/chefspec/default_spec.rb:19:in `block (3 levels) in <top (required)>'

Finished in 1.81 seconds (files took 1.23 seconds to load)
9 examples, 1 failure

Failed examples:

rspec ./test/chefspec/default_spec.rb:19 # sample-app::default on centos systems should install package "httpd"
```

I reverted Chefspec back to 4.1.1 in my Gemfile, but still the same issue. I heavily suspect that the Chef 11 -> 12 update is causing this.

Still not sure whether I am doing something wrong or whether this is indeed a bug...

**Comment by tknerr** [comment]
Just verified: downgrading Chef 11.16.4 resolves the issue, updating to Chef 12.3.0.dev.0 causes it.

**Comment by sethvargo** [comment]
So, Chef issue?

**Comment by tknerr** [comment]
@sethvargo not sure yet. Can only say: "in this specific case the package matcher works with Chef 11.16.4 but does not with 12.3.0.dev.0"

Will gather some more evidence...

And hey, good luck (though you probably don't need it -- so better: HF ;-)) for your talk at Philly ETE today! :

**Comment by tknerr** [comment]
[Added some logging](https://github.com/tknerr/sample-toplevel-cookbook/commit/10da9ecef2380597fbccd56de79961293ebb0559) to the apache2 default recipe. Looks like the attributes are being passed correctly (I just logged ERROR level so it stands out).

This is with Chef 11.16.4:
```
  on centos systems
[2015-04-07T05:35:14+00:00] ERROR: Hey, I'm running on centos and about to install httpd
    should install package "httpd"
[2015-04-07T05:35:14+00:00] ERROR: Hey, I'm running on centos and about to install httpd
    should start service "httpd"
```

And with Chef 12.3.0.dev.0:
```
  on centos systems
[2015-04-07T05:41:23+00:00] ERROR: Hey, I'm running on centos and about to install httpd
    should install package "httpd" (FAILED - 1)
[2015-04-07T05:41:23+00:00] ERROR: Hey, I'm running on centos and about to install httpd
    should start service "httpd"
```

So the `node.platform` and the computed `node['apache2']['package']` attributes seem to be properly passed to the recipe.

=> so it's not a Fauxhai or computed attributes problem

**Comment by tknerr** [comment]
Tried with latest released Chef version 12.2.1 as well, and the test fails here too:
```
  on centos systems
[2015-04-07T05:48:30+00:00] ERROR: Hey, I'm running on centos and about to install httpd
    should install package "httpd" (FAILED - 1)
[2015-04-07T05:48:30+00:00] ERROR: Hey, I'm running on centos and about to install httpd
    should start service "httpd"
```

=> not a newly introduced Chef master issue
=> more likely to be a ChefSpec / Chef 12.x incompatibility

**Comment by tknerr** [comment]
Weird. Tried a super simple example with the package_name attribute:
```ruby
package 'foo' do
  package_name 'httpd'
end
```

But the test for it still fails with Chef 12.2.1:
```ruby
  context "installing packages" do
    subject { ChefSpec::SoloRunner.converge(described_recipe) }
    it { should install_package 'httpd' }
  end
```

Output:
```
  installing packages
    should install package "httpd" (FAILED - 1)
    should install package "foo"
```

Does [this package identity example](https://github.com/sethvargo/chefspec/blob/8e496bd28b59cfc1d3f169f421265d03341b4fe7/examples/package/spec/install_spec.rb#L21) run against Chef 12.2.1, too?

**Comment by tknerr** [comment]
There must be something else to it. 

Your [package identity example](https://github.com/sethvargo/chefspec/blob/8e496bd28b59cfc1d3f169f421265d03341b4fe7/examples/package/spec/install_spec.rb#L21) runs fine, tried with Chef 11.16.4, 12.0.3 and 12.2.1 :+1: 

Need to dig deeper...

**Comment by tknerr** [comment]
Gotcha! Passing the fauxhai attributes makes it break (as of Chef 12.x).

For example, given this resource declaration:
```
package 'foo' do
  package_name 'httpd'
end
```

This spec works:
```
context "installing packages" do
    subject { ChefSpec::SoloRunner.converge(described_recipe) }
    it { should install_package 'httpd' }
    it { should install_package 'foo' }
  end
``` 

But this does not:
```
context "installing packages" do
    subject { ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '12.04').converge(described_recipe) }
    it { should install_package 'httpd' }
    it { should install_package 'foo' }
  end
```

I created a sample repro case here:
https://github.com/tknerr/sample-toplevel-cookbook/blob/chefspec-582/README.md

This is the commit which makes it break:
https://github.com/tknerr/sample-toplevel-cookbook/commit/6e83bf815af70a7cd78e3caaf1a622426822e207

**Comment by tknerr** [comment]
One could also argue, whether `it { should install_package 'foo' }` should pass in the above example where the identity attribute says the package "htop" is to be installed, but that would be a different issue

**Comment by tknerr** [comment]
found the root cause and pushed a fix to #583
